assets/js/components/Dashboard/NewPaMessage/StationsAndZones/SelectZonesPage.tsx

```diff
@@ -0,0 +1,541 @@
+      default:
+        return {
+          left: "Southbound",
+          middle: "Southbound & Northbound",
+          right: "Northbound",
+        };
+    }
+  }, [selectedRouteFilter]);
+
+  const massSelectButtonLabels = useMemo(() => {
+    if (selectedRouteFilter === "Silver") {
+      return {
+        left: "Both",
+        middle: "South Station/Downtown",
+        right: "Outbound Content",
+      };
+    }
+
+    return {
+      left: "All Zones (except bus)",
+      middle: directionLabels.left,
+      right: directionLabels.right,
+    };
+  }, [selectedRouteFilter]);
+
+  useEffect(() => {
+    const routes = fp.flow(
+      fp.flatMap((place: Place) =>
+        place.screens.filter((s) => value.includes(s.id)),
+      ),
+      fp.flatMap((screen: Screen) => screen.route_ids),
+      fp.uniq,
+      fp.groupBy((routeID: string) => {
+        if (routeID.startsWith("Green")) {
+          return "Green";
+        }
+
+        if (routeToRouteIDMap["Bus"].includes(routeID)) {
+          return "Bus";
+        }
+
+        if (routeToRouteIDMap["Silver"].includes(routeID)) {
+          return "Silver";
+        }
+
+        return routeID;
+      }),
+    )(places);
+
+    setSelectedRoutes(routes);
+    setSelectedRouteFilter(Object.keys(routes)[0]);
+    setNewPlaces(
+      places.filter((p) =>
+        places
+          .filter((p) => p.screens.some((s) => value.includes(s.id)))
+          .map((p) => p.id)
+          .includes(p.id),
+      ),
+    );
+  }, [places]);
+
+  const filteredPlaces = useMemo(() => {
+    return newPlaces.filter((p) =>
+      p.screens.some((s) =>
+        s.route_ids?.some((r) =>
+          routeToRouteIDMap[selectedRouteFilter]?.some((a) => r === a),
+        ),
+      ),
+    );
+  }, [selectedRouteFilter]);
+
+  const allScreens = useMemo(
+    () =>
+      filteredPlaces.flatMap((p) => {
+        return p.screens.filter(
+          (s) =>
+            fp.intersection(routeToRouteIDMap[selectedRouteFilter], s.route_ids)
+              .length > 0,
+        );
+      }),
+    [filteredPlaces],
+  );
+
+  const screensByZone = useMemo(() => {
+    const screensWithAZone = allScreens.filter((s) => s.zone);
+    return fp.groupBy((s) => s.zone, screensWithAZone);
+  }, [allScreens]);
+
+  const areAllGroupsSelected = (groups: Array<ReadonlyArray<string>>) =>
+    groups.every((group) =>
+      group.every((zone) => {
+        return signIDs(screensByZone[zone] ?? []).every(isSelected);
+      }),
+    );
+
+  const selectGroups = (groups: Array<ReadonlyArray<string>>) => {
+    const signsWithoutSelectedRoute = fp.without(signIDs(allScreens), value);
+
+    const newScreens = groups.flatMap((group) =>
+      group.flatMap((zone) => signIDs(screensByZone[zone] ?? [])),
+    );
+
+    onChange(fp.uniq([...signsWithoutSelectedRoute, ...newScreens]));
+  };
+
+  const isAllSelected =
+    value.length > 0 &&
+    areAllGroupsSelected([LEFT_ZONES, MIDDLE_ZONES, RIGHT_ZONES]);
```

Comment:
**suggestion:** if feels like the contents of `areAllGroupsSelected` could be in-lined here

Reply:
Maybe the function name is a bit misleading because "All" here refers to all parameters instead of all groups. I can rename for better clarity.